jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+"""
+Utilities for handling authentication in the Jobbergate system.
+"""
+import time
+from collections import namedtuple
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, cast
+
+import httpx
+from loguru import logger
+
+from .exceptions import AuthenticationError, TokenError
+from .token import Token, TokenType
+
+
+_LoginInformation = namedtuple(
+    "_LoginInformation",
+    ["verification_url", "wait_interval", "device_code"],
+)
+
+
+@dataclass
+class JobbergateAuth:
+    """
+    High-level class used to handle authentication in Jobbergate.
+
+    After an instance of this class is created, it can be used to authenticate requests
+    from the `requests`_ library or the `httpx`_ library by passing it to the ``auth``
+    parameter on the request (see examples bellow).
+
+    It works just out of the box. Behind the scenes, this procedure calls the
+    :meth:`JobbergateAuth.acquire_tokens` method to load the available tokens from the cache directory, it
+    will try to refresh them if they are expired, or will provide an URL to the user to login on the system.
+
+    Notice all steps above are also available individually as public methods, allowing a fine control for
+    advanced users.
+
+    .. _requests: https://requests.readthedocs.io/en/latest/
+    .. _httpx: https://www.python-httpx.org/
+
+    Arguments:
+        cache_directory (Path): Directory to be used for the caching tokens.
+        login_domain (str): Domain used for the login.
+        login_audience (str): Audience of the login.
+        login_client_id (str): Client ID used for login.
+        tokens (Dict[TokenType, Token]): Dictionary holding the tokens needed for authentication.
+
+    Note:
+        Consult the values above with your system administrator.
```

Comment:
I've written it more generically having in mind it is open-source documentation, and Jobbergate may be running on systems we do not have control.

At the same time, I see your point. Thoughts @matheushent @dusktreader @jamesbeedy?

Reply:
yes, I understand. I also thought in something like:

`
These values depend on the identity provider used for authentication. Consult your system administrator or contact the support for any guidance.
`

Reply:
it took my attention because seeing the documentation built, this text kind of jumps out
![image](https://user-images.githubusercontent.com/45343359/222160047-3926bff3-4bee-49cd-b2a1-65e2a0513807.png)

Reply:
> yes, I understand. I also thought in something like:
> 
> `These values depend on the identity provider used for authentication. Consult your system administrator or contact the support for any guidance.`

That is better, thanks. Let's see if we get more comments on this topic.

Reply:
The note was updated with your suggestion in [700ad32](https://github.com/omnivector-solutions/jobbergate/pull/228/commits/700ad32409f79be7a6483f32fa959954de32cd57)

Reply:
seems good to me, thank you

Reply:
Small correction,`These values depend on the identity provider used for authentication. Consult your system administrator or contact Omnivector support <support@omnivector.solutions> for further guidance. `

Reply:
Done